README.md
# [North American Numbering Plan Administrator](https://nationalnanpa.com/)

## NANPATelephoneFormatDetection

North American Numbering Plan 
The formatting rules:

- Easily Recognizable Codes	When the second and third digits of an area code are the same, that code is called an easily recognizable code (ERC). ERCs designate special services; e.g., 888 for toll-free service.
- N11	These 8 ERCs, called service codes, are not used as area codes.
- N9X	The 80 codes in this format, called expansion codes, have been reserved for use during the period when the current 10-digit NANP number format undergoes expansion.
- 37X and 96X	Two blocks of 10 codes each have been set aside by the INC for unanticipated purposes where it may be important to have a full range of 10 contiguous codes available.

- (?\<ERC\>^(\[2-9\](11\|22\|33\|44\|55\|66\|77\|88\|99)$))\|
  - (
    - (?\<NonGeoSvc\>^(5\(00\|21\|22\|23\|24\|25\|26\|27\|28\|29\|33\|44\|66\|77\|88\))\|
    - (?\<CAservices\>^(6(00\|22))\|
    - (?\<InterExch\>^(700))\|
    - (?\<USgovt\>^(710))|
    - (?\<TollFree\>^(8\(00\|33\|44\|55\|66\|77\|88)))\|
    - (?\<Premium\>^(900)
  - )(?\<SpecExchange\>(\[2-9\]\[0-9\]\[0-9\])-(?\<SpecNumber\>\[0-9\]{4}))
- (?\<ReservedAreaCode\>(\[2-9\]9\[0-9\])\|((37\|96)\[0-9\]))-\[2-9\]\[0-9\]\[0-9\]-\[0-9\]\{4\}$
- (?\<InfoAreaCode\>^(\[2-9\]\[0-8\]\[0-9\])-555-(?\<InfoNum\>(1212|4334)$)\|
- (?\<EntAreaCode\>^\[2-9\]\[0-8\]\[0-9\])-555-(?\<EntNumber>(01\[0-9\]\[0-9\]))$\|
- (?\<AreaCode\>(\[2-9\]\[0-8\]\[0-9\]))-(?\<Exchange>(\[2-9\]\[0-9\]\[0-9\]))-(?\<Number\>(\[0-9\]\{4\}))$

The rules:

1. Numbers must be in the following format: ddd, or ddd-ddd-dddd. No parentheses are allowed. 
2. If ERC is a group that is matched and the length is greater than 3, the number is invalid. 

[Zip file of Central Office Codes Report — Updated daily](https://www.nationalnanpa.com/nanp1/allutlzd.zip)



CHANGELOG.md
# Changelog

## [1.0.10](https://github.com/RalphHightower/NANPATelephoneFormatDetection/compare/v1.0.9...v1.0.10) (2023-06-11)


### Bug Fixes

* GitHub ([e6bfd92](https://github.com/RalphHightower/NANPATelephoneFormatDetection/commit/e6bfd92b97b1fa77dd7c70586db080e46b508259))

## [1.0.9](https://github.com/RalphHightower/NANPATelephoneFormatDetection/compare/v1.0.8...v1.0.9) (2023-02-27)


### Bug Fixes

* add 'or' to 900 for area code  ([520d2c3](https://github.com/RalphHightower/NANPATelephoneFormatDetection/commit/520d2c324ec41516fc5e8b246dbf37e33e03a4b3))
* balance parentheses ([e608fd5](https://github.com/RalphHightower/NANPATelephoneFormatDetection/commit/e608fd516e78ce8440d9079031bfd481ef45a75e))
* regular expression name group: LineNumber555 ([a1863e8](https://github.com/RalphHightower/NANPATelephoneFormatDetection/commit/a1863e8349f24919b42a0998e8e070f68fa245fe))

## [1.0.8](https://github.com/RalphHightower/NANPATelephoneFormatDetection/compare/v1.0.7...v1.0.8) (2023-02-13)


### Bug Fixes

* one area code didn’t have the 'or' ([ccd6603](https://github.com/RalphHightower/NANPATelephoneFormatDetection/commit/ccd660316b308dce6d9ffa7241dc24b385f38101))

## [1.0.7](https://github.com/RalphHightower/NANPATelephoneFormatDetection/compare/v1.0.6...v1.0.7) (2023-02-12)


### Bug Fixes

* ERC |66|66| to |66|77| ([be1ee8e](https://github.com/RalphHightower/NANPATelephoneFormatDetection/commit/be1ee8e26f7a80dba7dadf97b4d4b185aefa6694))

## [1.0.6](https://github.com/RalphHightower/NANPATelephoneFormatDetection/compare/v1.0.5...v1.0.6) (2023-01-29)


### Bug Fixes

* closes [#37](https://github.com/RalphHightower/NANPATelephoneFormatDetection/issues/37) ([059ebc2](https://github.com/RalphHightower/NANPATelephoneFormatDetection/commit/059ebc250c7d63b6401dabe57c3411461d85a4ed))

## [1.0.5](https://github.com/RalphHightower/NANPATelephoneFormatDetection/compare/v1.0.4...v1.0.5) (2023-01-26)


### Bug Fixes

* delete jekyll.yml (closes [#33](https://github.com/RalphHightower/NANPATelephoneFormatDetection/issues/33)) ([0b5c31d](https://github.com/RalphHightower/NANPATelephoneFormatDetection/commit/0b5c31d205e2697fd1b1c05917dd3438856868cc))

## [1.0.4](https://github.com/RalphHightower/NANPATelephoneFormatDetection/compare/v1.0.3...v1.0.4) (2023-01-26)


### Bug Fixes

* 866 toll free (closes [#26](https://github.com/RalphHightower/NANPATelephoneFormatDetection/issues/26)) ([d74428f](https://github.com/RalphHightower/NANPATelephoneFormatDetection/commit/d74428f391a021dd72923f91385cb2a4f3def530))

## [1.0.3](https://github.com/RalphHightower/NANPATelephoneFormatDetection/compare/v1.0.2...v1.0.3) (2023-01-22)


### Bug Fixes

* Sequence in parentheses ([09bb3bb](https://github.com/RalphHightower/NANPATelephoneFormatDetection/commit/09bb3bb37454af497904fbcee49659b8ffc0a43b))

## [1.0.2](https://github.com/RalphHightower/NANPATelephoneFormatDetection/compare/v1.0.1...v1.0.2) (2023-01-21)


### Bug Fixes

* string concatenation missing '+' (closes [#19](https://github.com/RalphHightower/NANPATelephoneFormatDetection/issues/19)) ([0de3c21](https://github.com/RalphHightower/NANPATelephoneFormatDetection/commit/0de3c2148233aaf426f5fca3bfb3566952aa5e9c))

## [1.0.1](https://github.com/RalphHightower/NANPATelephoneFormatDetection/compare/v1.0.0...v1.0.1) (2022-12-27)


### Bug Fixes

* various parentheses mismatches ([8d83699](https://github.com/RalphHightower/NANPATelephoneFormatDetection/commit/8d83699333e5433dbbf00426b2481d88c0e401cc))

## 1.0.0 (2022-12-27)


### Bug Fixes

* Add backslashes to brackets ([f60e9f1](https://github.com/RalphHightower/NANPATelephoneFormatDetection/commit/f60e9f1a254c01ba141e65050c987804901fd8dc))


MonsterRegularExpression_NANPA_formats.cs
#include System.Text.RegularExpressions;
///# [North American Numbering Plan Administrator](https://nationalnanpa.com/)
///
///- Easily Recognizable Codes	When the second and third digits of an area code are the same, that code is called an
///easily recognizable code (ERC). ERCs designate special services; e.g., 888 for toll-free service.
///- N11	These 8 ERCs, called service codes, are not used as area codes.
///- N9X	The 80 codes in this format, called expansion codes, have been reserved for use during the period when the 
///current 10-digit NANP number format undergoes expansion.
///- 37X and 96X	Two blocks of 10 codes each have been set aside by the INC for unanticipated purposes where it may be
///important to have a full range of 10 contiguous codes available.

namespace NorthAmericaNumberingPlanAdmistration
class TelephoneFormatDetection
{

    string TelephoneFormatRegularExpression = 
        "(^" +
            "(?<ERC>([2-79](11|22|33|44|55|66|77|88|99)))$|" +
            "(" +
                "(?<ReservedAreaCode>[2-9]9[0-9]|39[0-9]|96[0-9])|" +
                "(?<NonGeoSvc>(5[00|2[1-9]|33|44|66|77|88]))|" +
                "(?<CAservices>(6[00|22]))|" +
                "(?<InterExch>(700))|" +
                "(?<USgovt>(710))|" +
                "(?<TollFree>(8[00|33|44|55|66|77|88]))|" +
                "(?<Premium>(900))|" +
                "(?<AreaCode>([2-9][0-8][0-9]))" +
            ")-" +
            "((?<Exchange555>555)-(?<EntertainmentLineNumber>(01[0-9]{1}))|(?<DirectoryAssistance>(1212|4334))|((?<LineNumber555>)(0[2-9]{2}|1[0-9]{3})))|((?<Exchange>[2-9][0-9][0-9])-(?<LineNumber>[0-9{3}])))$";
//(^(?<ERC>([2-79](11|22|33|44|55|66|77|88|99)))$|((?<ReservedAreaCode>[2-9]9[0-9]|39[0-9]|96[0-9])|(?<NonGeoSvc>(5[00|2[1-9]|33|44|66|77|88]))|(?<CAservices>(6[00|22]))|(?<InterExch>(700))|(?<USgovt>(710))|(?<TollFree>(8[00|33|44|55|66|77|88]))|(?<Premium>(900))|(?<AreaCode>([2-9][0-8][0-9])))-((?<Exchange555>555)-(?<EntertainmentLineNumber>(01[0-9]{1}))|(?<DirectoryAssistance>(1212|4334))|((?<LineNumber555>)(0[2-9]{2}|1[0-9]{3})))|((?<Exchange>[2-9][0-9][0-9])-(?<LineNumber>[0-9{3}])))$";

    string test = "(^(?<ERC>([2-79](11|22|33|44|55|66|77|88|99)))$|((?<ReservedAreaCode>[2-9]9[0-9]|39[0-9]|96[0-9])|(?<NonGeoSvc>(5[00|2[1-9]|33|44|66|77|88]))|(?<CAservices>(6[00|22]))|(?<InterExch>(700))|(?<USgovt>(710))|(?<TollFree>(8[00|33|44|55|66|77|88]))|(?<Premium>(900))|(?<AreaCode>([2-9][0-8][0-9])))-((?<Exchange555>555)-(?<EntertainmentLineNumber>(01[0-9]{1}))|(?<DirectoryAssistance>(1212|4334))|((?<LineNumber555>)(0[2-9]{2}|1[0-9]{3})))|((?<Exchange>[2-9][0-9][0-9])-(?<LineNumber>[0-9{3}])))$";
}

